### PR TITLE
Improve query() function

### PR DESCRIPTION
**Pull request by joshflchan**
- Remove unnecessary logic to check and merge overlapping and conflicting busy times
- Improve query() function by using last finish time of interval
- Refactor code for readability

```
jchan@cloudshell:~/software-product-sprint/walkthroughs/week-4-tdd/project$ mvn -Dtest=FindMeetingQueryTest test
[INFO] Scanning for projects...
[INFO]
[INFO] --------------< com.google.sps:gcloud-tutorials-calendar >--------------
[INFO] Building gcloud-tutorials-calendar 1
[INFO] --------------------------------[ war ]---------------------------------
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ gcloud-tutorials-calendar ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/jchan/software-product-sprint/walkthroughs/week-4-tdd/project/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ gcloud-tutorials-calendar ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ gcloud-tutorials-calendar ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/jchan/software-product-sprint/walkthroughs/week-4-tdd/project/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ gcloud-tutorials-calendar ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ gcloud-tutorials-calendar ---
[INFO] Surefire report directory: /home/jchan/software-product-sprint/walkthroughs/week-4-tdd/project/target/surefire-reports
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.google.sps.FindMeetingQueryTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.209 sec
Results :
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.719 s
[INFO] Finished at: 2020-07-09T01:08:57-07:00
[INFO] ------------------------------------------------------------------------
jchan@cloudshell:~/software-product-sprint/walkthroughs/week-4-tdd/project$
```

### CODE REVIEW

**Comment by joshflchan** [comment]
This was quite satisfying! I had been revisiting the first PR every few days since it was reviewed. When first implementing the `query()` function, I definitely didn't feel too great about the idea of modifying and removing the elements of the `busyTimes` list. Thanks the explanation and pointing that out @KevinWorkman!

**Comment by joshflchan** [comment]
This was quite satisfying! I revisited the first PR a few time since after it was first reviewed. I had also tried using a ListIterator as suggested, but still couldn't seem to wrap my head around nesting the iterators.

When first implementing the query() function, I definitely didn't feel too great about the idea of modifying and removing the elements of the busyTimes list. Thanks for the explanation and pointing that out @KevinWorkman! I'm glad I revisited the problem again.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -24,38 +24,34 @@
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     ArrayList<TimeRange> freeTimes = new ArrayList<>();
-    if (request.getAttendees().isEmpty()){
-      return Arrays.asList(TimeRange.WHOLE_DAY);
-    }
+    ArrayList<TimeRange> busyTimes = getBusyTimes(events, request);
+    int lastFinish = TimeRange.START_OF_DAY;
+
     if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
```
Comment (KevinWorkman) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:30]: I would probably move this to be above the calculation for busyTimes. This is called a "guard clause" and I think it's usually more readable to put these as soon as possible in the code. (I think the if statement in the previous version was in the right place!)

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -24,38 +24,34 @@
...
 
-    for (int i = 0; i < busyTimes.size(); i++){
-      TimeRange timeRange = busyTimes.get(i);
-      if (i == 0){
-        TimeRange firstTimeRange = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, timeRange.start(), false);
-        if (firstTimeRange.start() != firstTimeRange.end()){
-          freeTimes.add(firstTimeRange); 
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    for (TimeRange timeRange : busyTimes){
+        if (timeRange.start() > lastFinish){
```
Comment (KevinWorkman) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]: I would maybe consider adding comments above each of these if statements very briefly explaining how they work. Something like:

```
// Add the free time between meetings.
if (timeRange.start() > lastFinish){
...
// Only add free time that's long enough for the requested meeting.
if (freeTimeRange.duration() >= request.getDuration()){
...
// Don't update lastFinish for meetings contained by timeRange.
if (timeRange.end() >= lastFinish){
...
// Get free time after all other meetings end.
if (lastFinish < TimeRange.END_OF_DAY){
```

Also don't be afraid to use more vertical whitespace. That's very subjective, but I personally fine an extra newline can help separate chunks of code into more logical pieces.

**Pull request merged**